### PR TITLE
Unified preprocessing for masking and parcellating

### PR DESCRIPTION
**Pull request by pbarbarant**
Solves #89. Creates a common preprocessing backend to be used with `PairwiseAlignment` and `TemplateAlignment`. Automatically masks and parcellates any number of nifti images in the spirit of scikit-learn's `StandardScaler`. Suggestions for better class name are welcome!

Goals:

- [x] Compatibility with `PairwiseAlignment`
- [x] Compatibility with `TemplateAlignment`
- [x] Compatibility with  `fastsrm`
- [x] Docstrings
- [x] Unit tests

### CODE REVIEW

**Comment by bthirion** [comment]
Is it ready for review ?

**Comment by pbarbarant** [comment]
> Is it ready for review ?

Apart from some docstrings that need updating, I think it's ready for review.

File: fmralign/preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                self.mask = reduced_mask
+                self.masker_ = check_embedded_masker(self)
+                self.masker_.n_jobs = self.n_jobs
+                self.masker_.fit()
+                warnings.warn(
+                    "Mask used was bigger than clustering provided. "
+                    + "Its intersection with the clustering was used instead."
+                )
+
+    def _one_parcellation(self, imgs):
```
Comment (bthirion) [inline fmralign/preprocessing.py:88]: Please add 1-line docstrings for private methods, full dosctrings for public methods.

File: fmralign/tests/test_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+import pytest
+from joblib import Memory
+from nibabel.nifti1 import Nifti1Image
+
+from fmralign._utils import ParceledData
+from fmralign.preprocessing import Preprocessor
+from fmralign.tests.utils import random_niimg
+
+
+def test_init_default_params():
```
Comment (bthirion) [inline fmralign/tests/test_preprocessing.py:11]: 1-line docstrings for tests is very helpful for test mainetnance...

File: fmralign/tests/test_utils.py [hunk 3]
```diff
@@ -1,52 +1,194 @@
...
     with pytest.warns(UserWarning):
-        clustering = nibabel.Nifti1Image(
+        clustering = nib.Nifti1Image(
             np.hstack([np.ones(2000), 4 * np.ones(800)]), np.eye(4)
         )
-        _make_parcellation(img, indexes, clustering_method, n_pieces, masker)
+        _make_parcellation(img, clustering_method, n_pieces, masker)
+
+
+def test_initialization():
```
Comment (bthirion) [inline fmralign/tests/test_utils.py:53]: 1-line docsrtings welcome too...

**Review by bthirion — COMMENTED** [review]
Still wondering about the necessity of the class (to be discussed)/
Docstrings should be a bit richer.

**Review by bthirion — COMMENTED** [review]
there is one comment pending, LGTM otherwise.

**Comment by pbarbarant** [comment]
> there is one comment pending, LGTM otherwise.

I'm still not super satisfied with the class name `Preprocessor` even though it's an internal class. @bthirion @emdupre would `ParcellationMasker` cut it ? It needs to be distinct enough from `NiftiLabelsMasker`.

**Comment by bthirion** [comment]
> > there is one comment pending, LGTM otherwise.
> 
> I'm still not super satisfied with the class name `Preprocessor` even though it's an internal class. @bthirion @emdupre would `ParcellationMasker` cut it ? It needs to be distinct enough from `NiftiLabelsMasker`.

Happy with `ParcellationMasker`

**Comment by pbarbarant** [comment]
Waiting for #105 to be merged and then merging it.

**Comment by emdupre** [comment]
Thank you, @pbarbarant ! Happy to see this in 🚀

**Pull request merged**